src/Ngsa.Middleware/NgsaLog.cs

```diff
@@ -12,142 +12,161 @@ namespace Ngsa.Middleware
 
-                counter++;
+                Console.ForegroundColor = ConsoleColor.Yellow;
+                Console.WriteLine(JsonSerializer.Serialize(d, Options));
+                Console.ResetColor();
             }
-
-            // todo - reuse loggers
-            // use iDisposable?
-            //Loggers.Add(counter, logger);
-
-            return logger;
         }
 
-        public void LogInformation(string message)
+        /// <summary>
+        /// Log warning
+        /// </summary>
+        /// <param name="eventId">Event ID</param>
+        /// <param name="method">method to log</param>
+        /// <param name="message">message to log</param>
+        /// <param name="context">http context</param>
+        public void LogWarning(EventId eventId, string method, string message, HttpContext context = null)
         {
-            const LogLevel logLevel = LogLevel.Information;
-
-            if (LogLevel <= logLevel)
+            if (LogLevel >= LogLevel.Warning)
             {
-                Dictionary<string, object> d = GetDictionary(message, logLevel);
+                Dictionary<string, object> d = GetDictionary(eventId, method, message, LogLevel.Warning, context);
 
-                Console.ForegroundColor = ConsoleColor.Green;
+                Console.ForegroundColor = ConsoleColor.Yellow;
                 Console.WriteLine(JsonSerializer.Serialize(d, Options));
                 Console.ResetColor();
             }
         }
 
-        public void LogWarning(string message)
+        /// <summary>
+        /// Log error
+        /// </summary>
+        /// <param name="eventId">Event ID</param>
+        /// <param name="method">method to log</param>
+        /// <param name="message">message to log</param>
+        /// <param name="context">http context</param>
+        /// <param name="ex">exception</param>
+        public void LogError(EventId eventId, string method, string message, HttpContext context = null, Exception ex = null)
         {
-            const LogLevel logLevel = LogLevel.Warning;
-
-            if (LogLevel > logLevel)
+            if (LogLevel >= LogLevel.Error)
             {
-                return;
-            }
+                Dictionary<string, object> d = GetDictionary(eventId, method, message, LogLevel.Error, context);
 
-            Dictionary<string, object> d = GetDictionary(message, logLevel);
+                if (ex != null)
+                {
+                    d.Add("ExceptionType", ex.GetType().FullName);
+                    d.Add("ExceptionMessage", ex.Message);
+                }
 
-            Console.ForegroundColor = ConsoleColor.Yellow;
-            Console.WriteLine(JsonSerializer.Serialize(d, Options));
-            Console.ResetColor();
+                // display the error
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.Error.WriteLine(JsonSerializer.Serialize(d, Options));
+                Console.ResetColor();
```

Comment:
feels like we should move the above 3 lines to a common function and have each login function call it with something like:
`WriteLogEntry(ConsoleColor.Red, d)`

Thoughts?

Reply:
changed per code review